File: JavaScript/Easy/MeetingRooms.js
/**
 * @param {number[][]} intervals
 * @return {boolean}
 */
var canAttendMeetings = function(intervals) {
  
  for (let i = 0; i < intervals.length; i++) {
    for (let j = i + 1; j < intervals.length; j++) {
      let minEnd = Math.min(intervals[i][1], intervals[j][1]);
      let maxStart = Math.max(intervals[i][0], intervals[j][0]);
      if (minEnd > maxStart)
        return false;
    }
  }
  
  return true;
};

/*

  Approach 1:
  
  Brute force nested for loop
  
  Check for overlap between two non-sorted intervals by comparing the minimum end time with the max start time
  
  Time: O(n^2)
  Space: O(1)

*/



/**
 * @param {number[][]} intervals
 * @return {boolean}
 */
var canAttendMeetings = function(intervals) {
  
  intervals.sort((a,b) => a[0] - b[0]);
  
  for (let i = 1; i < intervals.length; i++) {
    let interval1 = intervals[i - 1];
    let interval2 = intervals[i];
    if (interval1[1] > interval2[0])
      return false;
  }
    
  return true;
};

/*

  Approach 2:

  Sort based on start value and check if the end value exceeds the start of the next interval

  Time: O(nlogn)
  Space: O(1)

*/

File: C#/Easy/MeetingRooms.cs
public class Solution {
    public bool CanAttendMeetings(int[][] intervals)
    {
        if (intervals.Length == 0) return true;

        Array.Sort(intervals, (a, b) => a[0].CompareTo(b[0]));
        for (int i = 1; i < intervals.Length; i++)
        {
            int[] interval1 = intervals[i - 1];
            int[] interval2 = intervals[i];
            if (interval1[1] > interval2[0]) return false;
        }

        return true;
    }
}

/*

    Example 1
    Input = [[0, 5], [5, 10], [11, 14]]
    Output = true

    Example 2
    Input = [[0, 5], [5, 10], [7, 11]]
    Output = false (because theres overlap between intervals[1] and intervals[2])

    Example 3
    Input = []
    Output = true

    Solution 1
    - Compare all combination of meetings (MeetingA and MeetingB)
    - Return false if overlap is found, otherwise return true
    Time: O(n^2)
    Space: O(1)
    
    Optimization
    - Sort intervals by start time
    - Compare end time interval[i] with interval[i + 1]
    - If the end time > start time, return false, otherwise return true
    Time: O(nlogn)
    Space: O(1) (depending on sorting algorithm)

*/